### PR TITLE
feat: add eslint plugin no-html-links

### PR DESCRIPTION
**Pull request by JohnVicke**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.

## Motivation
Described in:  https://github.com/facebook/docusaurus/issues/6472

## Test Plan
Jest unit tests are included. 

<details>
<summary>Output of `yarn test no-html-plugins` </summary>
<br>

```
PASS packages/eslint-plugin/src/rules/__tests__/no-html-links.test.ts
  prefer-docusaurus-link
    valid
      ✓ <Link to="/test">test</Link> (239 ms)
      ✓ <Link to="https://twitter.com/docusaurus">Twitter</Link> (2 ms)
    invalid
      ✓ <a href="/test">test</a> (3 ms)
      ✓ <a href="https://twitter.com/docusaurus" target="_blank">test</a> (2 ms)
      ✓ <a href="https://twitter.com/docusaurus" target="_blank" rel="noopener noreferrer">test</a> (1 ms)

Test Suites: 1 passed, 1 total
Tests:       5 passed, 5 total
Snapshots:   0 total
Time:        0.67 s, estimated 1 s
Ran all test suites matching /no-html-links/i.
```

</details>

### Test links

Docs preview: https://deploy-preview-8156--docusaurus-2.netlify.app/docs/api/misc/@docusaurus/eslint-plugin/no-html-links

### Comments
The solution is fairly simple, I played around with when the message should be returned to the user (e.g. a tag has no attributes). However, if we want to enforce docusaurus links all the time this feels more accurate.

### CODE REVIEW

**Review by slorber — CHANGES_REQUESTED** [review]
Thanks, LGTM 👍 

Wonder if we want to be that strict and never allow any absolute URL by default, so maybe adding an option can be convenient?

We mostly want to prevent bad relative links IMHO

Any opinion @Josh-Cena ?

**Comment by Josh-Cena** [comment]
> Wonder if we want to be that strict and never allow any absolute URL by default, so maybe adding an option can be convenient?

An option like `ignoreFullyResolved` would be my go-to, yes. It would be tricky to check template literal values, though. Recommending always using `<Link>` is not _that_ bad and may allow us to do other useful things in the future.

File: packages/docusaurus-theme-classic/src/theme/TOCItems/Tree.tsx
Comment (Josh-Cena) [reply packages/docusaurus-theme-classic/src/theme/TOCItems/Tree.tsx:27]: It's not undoable. Even for template literals, you just need to check `quasis[0][0] === "#"`.

**Comment by slorber** [comment]
`ignoreFullyResolved` seems fine, even if it only allows hardcoded href props that can already be convenient.

Not sure "fully resolved" is very common, where does it come from?

---

> Recommending always using <Link> is not that bad and may allow us to do other useful things in the future.

Agree, but maybe we'll need to improve it then because for example not merging `rel` (current behavior) is a bit annoying

File: website/docs/api/misc/eslint-plugin/no-html-links.md
Comment (JohnVicke) [reply website/docs/api/misc/eslint-plugin/no-html-links.md:24]: Do you think it would suffice with adding an extra line of correct/incorrect usage or something more descriptive? 

`<a href="https://wwww,twitter.com/docusaurus" target="_blank">Twitter</a>`

`<Link to="https://www.twitter.com/docusaurus">Twitter</Link>`

And maybe we should link ( :drum: ) to the Link docs from here as well?

**Comment by JohnVicke** [comment]
> ignoreFullyResolved seems fine, even if it only allows hardcoded href props that can already be convenient.

Maybe we also should add an option to ignore `mailto` or let them pass as default behaviour?

**Comment by Josh-Cena** [comment]
I'd say, anything that starts with a protocol should be ignored.

File: packages/docusaurus-theme-classic/src/theme/TOCItems/Tree.tsx
Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/TOCItems/Tree.tsx:27]: I guess we can just ignore hashes. This is a bonus, we can merge this PR without this feature if complicated to implement.

**Comment by Josh-Cena** [comment]
I was thinking about "fully resolved paths". I don't know where that term crosses my mind either.

File: website/docs/api/misc/eslint-plugin/no-html-links.md
Comment (slorber) [inline website/docs/api/misc/eslint-plugin/no-html-links.md:39]: if "fully resolved" === "absolute url" then maybe we'd want to be consistent?

https://www.rfc-editor.org/rfc/rfc3986#section-4.3

Maybe we just want "ignoreAbsoluteUrls" instead, it would be easier for users to understand?

This is just wording, but it remains important. 

Now we can still do a breaking change later if we change our mind, not a huge deal.

File: website/docs/api/misc/eslint-plugin/no-html-links.md
Comment (slorber) [reply website/docs/api/misc/eslint-plugin/no-html-links.md:43]: Yes, just add more examples of good/bad usage

And a link to Link 👍 why not, users could find this helpful

**Comment by Josh-Cena** [comment]
I want to review this before merging. I think I can find some time this weekend

**Comment by slorber** [comment]
conflict resolved

@Josh-Cena do you still want to review it?

**Comment by slorber** [comment]
@Josh-Cena ?

**Review by Josh-Cena — COMMENTED** [review]
Extremely sorry for missing this. The implementation looks good although I'd like to work on enhancements after this is merged. Just requesting a bit more tests (as ESLint plugin maintainers always do...)

**Comment by slorber** [comment]
@Josh-Cena let me know if you want to do something about this PR 🤪 otherwise I'll apply some suggestions, do a few minor changes and merge it

**Comment by Josh-Cena** [comment]
Mostly LGTM. I wonder if we should add an option to ignore fully dynamic and not statically analyzable `href` props, but that could be an extension because using `Link` can't harm in any case.

**Comment by slorber** [comment]
Look good enough to merge

Added some extra code to handle hardcoded/static template literal strings too `href={"https://xyz.com"}` as it wasn't the case.

For more dynamic cases where static analysis is possible, that can be handled in a follow-up PR

---

Was wondering if it's possible to pre-evaluate AST nodes with TSESTree @Josh-Cena @JoshuaKGoldberg 

Apparently, Babel is able to do that successfully:

CleanShot 2022-12-14 at 18 04 52@2x

Feel free to complete the missing cases if you know how to ;)

**Pull request merged**
